server/controllers/stock/lots.js

```diff
@@ -166,42 +167,124 @@ function getDupes(req, res, next) {
-      transaction.addQuery(updateLotTags, [uuid, mergeLotUuid]);
-      transaction.addQuery(updateStockAssign, [uuid, mergeLotUuid]);
-      transaction.addQuery(updateStockMovement, [uuid, mergeLotUuid]);
-      transaction.addQuery(deleteLot, [mergeLotUuid]);
-    });
-    await transaction.execute();
-    res.sendStatus(200);
-  } catch (error) {
-    next(error);
-  }
+  const transaction = db.transaction();
+
+  lotsToMerge.forEach(rawUuid => {
+    const mergeLotUuid = db.bid(rawUuid);
+    transaction.addQuery(updateLotTags, [keepLotUuid, mergeLotUuid]);
+    transaction.addQuery(updateStockAssign, [keepLotUuid, mergeLotUuid]);
+    transaction.addQuery(updateStockMovement, [keepLotUuid, mergeLotUuid]);
+    transaction.addQuery(deleteLot, [mergeLotUuid]);
+  });
+
+  return transaction.execute();
+}
+
+/**
+ * GET /lots/:uuid/merge
+ *
+ * @description
+ * Merge the lots_to_merge (in the body) into the lot to keep (given by uuid).
+ * This is a accomplished in two steps for each lot to merge:
+ *   1. Replace all references to the lot to be merged with
+ *      references to the lot to keep.
+ *   2. Delete the lot to be merged
+ */
+function merge(req, res, next) {
+  const { uuid } = req.params;
+  const lotsToMerge = req.body.lotsToMerge.map(db.bid);
+
+  mergeLotsInternal(uuid, lotsToMerge)
+    .then(() => {
+      res.sendStatus(200);
+    })
+    .catch(next)
+    .done();
+}
+
+/**
+ * GET /lots/merge/auto
+ *
+ * @description
+ * Finds and merges all lots suitable for automatic merging
+ *  - To qualify, the lots must have the same inventory_uuid,
+ *    label, and expiration date.
+ */
+function autoMerge(req, res, next) {
+  // The first query gets the inventory UUID for each
+  // inventory article with duplicate lots (having the
+  // same label, inventory_uuid, and expiration_date).
+  // (Since these are grouped, only one of the several
+  // lots is given.)
+  const query1 = `
+  SELECT
+    BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+    BUID(i.uuid) AS inventory_uuid, i.text as inventory_text,
+    COUNT(*) as num_duplicates
+  FROM lot l
+  JOIN inventory i ON i.uuid = l.inventory_uuid
+  GROUP BY label, inventory_uuid, expiration_date HAVING num_duplicates > 1
+`;
+  // The second query gets all lots matching the label,
+  // inventory_uuid, and expiration dates
+  const query2 = `
+    SELECT
+      BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+      BUID(i.uuid) AS inventory_uuid, i.text as inventory_text
+    FROM lot l
+    JOIN inventory i ON i.uuid = l.inventory_uuid
+    WHERE l.label=? AND i.uuid=? AND l.expiration_date=DATE(?)
+  `;
+  db.exec(query1, [])
+    .then((rows) => {
+      const numInventories = rows.length;
+      const numLots = rows.reduce((sum, row) => {
+        return sum + row.num_duplicates;
+      }, 0) - numInventories;
+      const dbPromises = [];
+      rows.forEach((row) => {
+        db.exec(query2, [row.label, db.bid(row.inventory_uuid), row.expiration_date])
```

Comment:
This `db.exec()` is asynchronous, but this code doesn't treat it as such.  The `row.forEach()` is going to just keep spinning without ensuring the `db.exec()`s are done, either in parallel or in serial.  You really should account for this.

You're close with your `dbPromises` approach.  However, because you aren't waiting for these `db.exec()`s to return, you may careen right past it without actually waiting.

A thought exercise: imagine that each `db.exec()` here takes 15 seconds.  What will happens is that `rows.forEach()` will churn through the array before a single `.then()` is called.  That means that `dbPromises` will be empty after running through the `.forEach()` and the `Promise.all()` call on an empty array will successfully return nothing.

To prevent this, let's `return mergeLotsInternal(keepLotUuid, lotUuids))` at the end of the `.then()` call.  What does this do?  It links the `mergeLotsInternal` call into the promise chain.  To `node` it will look like you've just added another `.then()` to the chain.  After, let's push the _entire chain_ into the `dbPromises` array.  Effectively:
```js
rows.forEach(row => {
  const promise = db.exec(/* params */)
    .then((lots) => {
       // do all kinds of cool processing here
       return mergeLotsInternal(/* more params */);
    });
    
  // add the entire chain to the dbPromises array
  dbPromises.push(promise);
});

// now when we call Promise.all(), we are waiting for the entire promise chain to complete.  That means
// the db.exec() and every .then() after it, including the mergeLotsInternal() function
Promise.all(dbPromises)
  .then(() => { /* return http status code */ })
  // remember error handling!  This will handle errors for the entirety of the chain.
  .catch(err => next(err));
```

Reply:
Fixed with your suggestion.  Thanks!